File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/Queue/Recursion/QueueRecursion.py
# Recursion in Queue - Recursively dequeue until empty (MotorPH)


def enqueue(queue, brand, quantity):
    """Enqueue inventory process (add to back)."""
    queue.append({"brand": brand, "quantity": quantity})


def recursive_dequeue(queue):
    """Recursively dequeue queue until empty (base case: empty queue)."""
    if not queue:
        return
    item = queue.pop(0)
    print(f"  Dequeued: {item['brand']}: {item['quantity']} units")
    recursive_dequeue(queue)


def display(queue):
    """Display queue (front to back)."""
    if not queue:
        print("  (queue empty)")
        return
    for i, item in enumerate(queue):
        print(f"  {i + 1}. {item['brand']}: {item['quantity']} units")


def main():
    # Queue for inventory process
    queue = []

    # Enqueue inventory process
    enqueue(queue, "Honda", 5)
    enqueue(queue, "Yamaha", 4)
    enqueue(queue, "Suzuki", 7)

    print("Queue before recursive dequeue:")
    display(queue)

    # Recursively dequeue until empty
    print("\nRecursive dequeue (emptying queue):")
    recursive_dequeue(queue)

    print("\nQueue after recursive dequeue:")
    display(queue)


if __name__ == "__main__":
    main()
#

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/Queue/Recursion/QueueRecursion.java
package Queue.Recursion;

import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Recursion in Queue - Recursively dequeue until empty (MotorPH)
 */
public class QueueRecursion {

    record QueueItem(String brand, int quantity) {}

    /** Enqueue inventory process (add to back). */
    static void enqueue(Deque<QueueItem> queue, String brand, int quantity) {
        queue.addLast(new QueueItem(brand, quantity));
    }

    /** Recursively dequeue queue until empty (base case: empty queue). */
    static void recursiveDequeue(Deque<QueueItem> queue) {
        if (queue.isEmpty()) {
            return;
        }
        QueueItem item = queue.removeFirst();
        System.out.println("  Dequeued: " + item.brand() + ": " + item.quantity() + " units");
        recursiveDequeue(queue);
    }

    /** Display queue (front to back). */
    static void display(Deque<QueueItem> queue) {
        if (queue.isEmpty()) {
            System.out.println("  (queue empty)");
            return;
        }
        int i = 1;
        for (QueueItem item : queue) {
            System.out.println("  " + i++ + ". " + item.brand() + ": " + item.quantity() + " units");
        }
    }

    public static void main(String[] args) {
        // Queue for inventory process (FIFO - ArrayDeque: addLast = enqueue, removeFirst = dequeue)
        Deque<QueueItem> queue = new ArrayDeque<>();

        // Enqueue inventory process
        enqueue(queue, "Honda", 5);
        enqueue(queue, "Yamaha", 4);
        enqueue(queue, "Suzuki", 7);

        System.out.println("Queue before recursive dequeue:");
        display(queue);

        // Recursively dequeue until empty
        System.out.println("\nRecursive dequeue (emptying queue):");
        recursiveDequeue(queue);

        System.out.println("\nQueue after recursive dequeue:");
        display(queue);
    }
}
